pages/api/user/index.ts

```diff
@@ -0,0 +1,180 @@
+}
+
+async function handlePostWithOldToken(
+  currentUid: string,
+  oldUserToken: string,
+  req: NextApiRequest,
+  res: NextApiResponse<ApiResponse<EmptyPayload>>,
+) {
+  const currentIsAnonymous = (await getUserFromJwt(req.headers.authorization as string)).isAnonymous;
+  if (currentIsAnonymous) {
+    return res
+      .status(HttpStatus.UNAUTHORIZED)
+      .json(createJsonResponse({}, { type: ERROR, message: 'An unverified account cannot be the target of link.' }));
+  }
+
+  let oldUserDetails: UserDetailsFromRequest;
+  try {
+    oldUserDetails = await getUserFromJwt(oldUserToken);
+  } catch (error) {
+    return res
+      .status(HttpStatus.UNAUTHORIZED)
+      .json(createJsonResponse({}, { type: ERROR, message: `Could not validate and decode 'old_token'.` }));
+  }
+
+  const oldUid = oldUserDetails.uid;
+  if (!oldUserDetails.isAnonymous) {
+    return res
+      .status(HttpStatus.BAD_REQUEST)
+      .json(createJsonResponse({}, { type: ERROR, message: `UID ${oldUid} is already linked` }));
+  }
+
+  const result = await linkAccount(oldUid, currentUid);
+
+  switch (result) {
+    case LinkAccountResult.OLD_USER_DELETED:
+      return res.status(HttpStatus.OK).json(
+        createJsonResponse(
+          {},
+          {
+            type: INFORMATION,
+            message: `Old UID ${oldUid} is no longer in database. New user ${currentUid} has been linked.`,
+          },
+        ),
+      );
+
+    case LinkAccountResult.NEW_USER_HAS_DATA:
+      return res
+        .status(HttpStatus.CONFLICT)
+        .json(createJsonResponse({}, { type: ERROR, message: 'New UID already has user data and cannot be linked.' }));
+
+    case LinkAccountResult.NEW_USER_LINKED:
+      return res
+        .status(HttpStatus.OK)
+        .json(
+          createJsonResponse(
+            {},
+            { type: SUCCESS, message: `Old UID ${oldUid} has been replaced with new UID ${currentUid}.` },
+          ),
+        );
+  }
+}
+
+async function handleDelete(uid: string, _req: NextApiRequest, res: NextApiResponse<ApiResponse<EmptyPayload>>) {
+  // UID is a primary key, so the count will either be 0 (UID does not exist in database) or 1 (successful delete).
+  const user = await getUserIfExists(uid);
+  if (!user) {
+    return res
+      .status(HttpStatus.NOT_FOUND)
+      .json(createJsonResponse({}, { type: ERROR, message: `Could not find UID ${uid}.` }));
+  }
+
+  await prisma.application.deleteMany({ where: { userId: uid } });
+  await prisma.user.delete({ where: { uid } });
+
+  return res
+    .status(HttpStatus.OK)
+    .json(createJsonResponse({}, { type: SUCCESS, message: `Deleted UID ${uid} and their data.` }));
+}
+
+async function linkAccount(oldUid: string, newUid: string): Promise<LinkAccountResult> {
+  const newUser = await getUserIfExists(newUid);
+  const oldUser = await getUserIfExists(oldUid);
+
+  // If old user has already been removed from the database. This can happen if the API was called twice or if the new user
```

Comment:
I thought if `oldUid` is found and `verifiedUid` is not found, we are updating `oldUid` -> `verifiedUid`, without creating or deleting any old/new user?

I think this means the other cases are automatically handled.

This would allow cascading updates to happen so we can be assured the rest of our relations are correct?

Reply:
Yes. I am also confused about this. During our discussion, I thought we agreed on just updating the anon id -> verified id when user logs in with github. Why would there be a need to delete from the db?

Reply:
if anonUid is updated to verifiedUid, the original anonUid is no longer in the database. I said "already removed from database" to indicate the old uid might have been deleted or already linked to some other, unknown verifiedUid.

Reply:
Resolved: we will never allow relinking.